Error migrating to the 50.0 with the database connection encrypted

### Describe the bug

I have a Metabase instance that runs in the docker container with the database connection encrypted with MB_ENCRYPTION_SECRET_KEY. 
During the upgrade to 50.0 version, database migration fails. I believe that's caused by the fill_cache_config.pg.sql migration script that tries to set up the default settings for the new cache config but doesn't take into the account the fact that this setting is encrypted.
I believe it might be an issue for the migrations on the other database engines as well.
The workaround is to remove the settings keys affected by the migration scripts before the migration.

### To Reproduce

1. Have the pre-0.50.0 Metabase instance setup with MB_ENCRYPTION_SECRET_KEY setting and the query caching set up.
2. Upgrade to 0.50.0 version.
3. See the error during migration.


### Expected behavior

Migration completes successfully.

### Logs

```
2024-06-11 15:56:56 2024-06-11 12:56:56,331 INFO db.liquibase :: Database has unrun migrations. Checking if migration lock is taken...
2024-06-11 15:56:56 2024-06-11 12:56:56,348 INFO db.liquibase :: No migration lock found.
2024-06-11 15:56:56 2024-06-11 12:56:56,348 INFO db.liquibase :: Migration lock acquired.
2024-06-11 15:56:56 2024-06-11 12:56:56,609 INFO db.liquibase :: Running 32 migrations ...
2024-06-11 15:56:56 2024-06-11 12:56:56,932 ERROR liquibase.changelog :: ChangeSet migrations/001_update_migrations.yaml::v50.2024-04-12T12:33:09::piranha encountered an exception.
2024-06-11 15:56:56 liquibase.exception.DatabaseException: ERROR: invalid input syntax for type double precision: "BW2J+jGXv2seSOh2ZXEehWTUWTwUJx5lSUr20Sw6mulcSaaFxjEmtgtLwE+wacV5CqyebuY/zg/e5pP7Wg5m3Q==" [Failed SQL: (0) WITH root AS (
2024-06-11 15:56:56   select 'root' AS model,
2024-06-11 15:56:56          0      AS model_id,
2024-06-11 15:56:56          'ttl'  AS strategy,
2024-06-11 15:56:56          json_build_object(
2024-06-11 15:56:56            'multiplier', coalesce((select value::float::int from setting where key = 'query-caching-ttl-ratio'), 10),
2024-06-11 15:56:56            'min_duration_ms', coalesce((select value::float::int from setting where key = 'query-caching-min-ttl'), 60000)
2024-06-11 15:56:56          ) AS config
2024-06-11 15:56:56 ), database AS (
2024-06-11 15:56:56   select 'database' AS model,
2024-06-11 15:56:56          id         AS model_id,
2024-06-11 15:56:56          'duration' AS strategy,
2024-06-11 15:56:56          json_build_object('duration', cache_ttl, 'unit', 'hours') AS config
2024-06-11 15:56:56     from metabase_database
2024-06-11 15:56:56    where cache_ttl is not null
2024-06-11 15:56:56 ), dashboard AS (
2024-06-11 15:56:56   select 'dashboard' AS model,
2024-06-11 15:56:56          id          AS model_id,
2024-06-11 15:56:56          'duration'  AS strategy,
2024-06-11 15:56:56          json_build_object('duration', cache_ttl, 'unit', 'hours') AS config
2024-06-11 15:56:56     from report_dashboard
2024-06-11 15:56:56    where cache_ttl is not null
2024-06-11 15:56:56 ), card AS (
2024-06-11 15:56:56   select 'question' AS model,
2024-06-11 15:56:56          id         AS model_id,
2024-06-11 15:56:56          'duration' AS strategy,
2024-06-11 15:56:56          json_build_object('duration', cache_ttl, 'unit', 'hours') AS config
2024-06-11 15:56:56     from report_card
2024-06-11 15:56:56    where cache_ttl is not null
2024-06-11 15:56:56 ), rows1 AS (
2024-06-11 15:56:56   SELECT * FROM root UNION ALL
2024-06-11 15:56:56   SELECT * FROM database UNION ALL
2024-06-11 15:56:56   SELECT * FROM dashboard UNION ALL
2024-06-11 15:56:56   SELECT * FROM card
2024-06-11 15:56:56 ), rows AS (
2024-06-11 15:56:56   SELECT * FROM rows1
2024-06-11 15:56:56    WHERE (SELECT true FROM setting WHERE key = 'enable-query-caching' AND VALUE = 'true')
2024-06-11 15:56:56 )
2024-06-11 15:56:56     INSERT INTO cache_config (model, model_id, strategy, config)
2024-06-11 15:56:56     SELECT * FROM rows
2024-06-11 15:56:56     ON CONFLICT (model, model_id) DO NOTHING]
2024-06-11 15:56:56     at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(jdbcexecutor.java:470)
2024-06-11 15:56:56     at liquibase.executor.jvm.JdbcExecutor.execute(jdbcexecutor.java:77)
2024-06-11 15:56:56     at liquibase.executor.jvm.JdbcExecutor.execute(jdbcexecutor.java:179)
2024-06-11 15:56:56     at liquibase.executor.AbstractExecutor.execute(abstractexecutor.java:141)
2024-06-11 15:56:56     at liquibase.database.AbstractJdbcDatabase.executeStatements(abstractjdbcdatabase.java:1285)
2024-06-11 15:56:56     at liquibase.changelog.ChangeSet.execute(changeset.java:755)
2024-06-11 15:56:56     at liquibase.changelog.visitor.UpdateVisitor.executeAcceptedChange(updatevisitor.java:119)
2024-06-11 15:56:56     at liquibase.changelog.visitor.UpdateVisitor.visit(updatevisitor.java:68)
2024-06-11 15:56:56     at liquibase.changelog.ChangeLogIterator$2.lambda$run$0(changelogiterator.java:133)
2024-06-11 15:56:56     at liquibase.changelog.ChangeLogIterator$2.lambda$run$0(changelogiterator.java:133)
2024-06-11 15:56:56     at liquibase.Scope.lambda$child$0(scope.java:186)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:195)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:185)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:164)
2024-06-11 15:56:56     at liquibase.changelog.ChangeLogIterator$[2.run](https://2.run/)(changelogiterator.java:122)
2024-06-11 15:56:56     at liquibase.Scope.lambda$child$0(scope.java:186)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:195)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:185)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:164)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:252)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:256)
2024-06-11 15:56:56     at [liquibase.changelog.ChangeLogIterator.run](https://liquibase.changelog.changelogiterator.run/)(changelogiterator.java:89)
2024-06-11 15:56:56     at liquibase.command.core.AbstractUpdateCommandStep.lambda$run$0(abstractupdatecommandstep.java:110)
2024-06-11 15:56:56     at liquibase.Scope.lambda$child$0(scope.java:186)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:195)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:185)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:164)
2024-06-11 15:56:56     at [liquibase.command.core.AbstractUpdateCommandStep.run](https://liquibase.command.core.abstractupdatecommandstep.run/)(abstractupdatecommandstep.java:108)
2024-06-11 15:56:56     at [liquibase.command.core.UpdateCommandStep.run](https://liquibase.command.core.updatecommandstep.run/)(updatecommandstep.java:105)
2024-06-11 15:56:56     at liquibase.command.CommandScope.execute(commandscope.java:217)
2024-06-11 15:56:56     at liquibase.Liquibase.lambda$update$0(liquibase.java:245)
2024-06-11 15:56:56     at liquibase.Scope.lambda$child$0(scope.java:186)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:195)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:185)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:164)
2024-06-11 15:56:56     at liquibase.Liquibase.runInScope(liquibase.java:1419)
2024-06-11 15:56:56     at liquibase.Liquibase.update(liquibase.java:234)
2024-06-11 15:56:56     at liquibase.Liquibase.update(liquibase.java:212)
2024-06-11 15:56:56     at liquibase.Liquibase.update(liquibase.java:194)
2024-06-11 15:56:56     at metabase.db.liquibase$migrate_up_if_needed_BANG_$fn__44378.invoke(liquibase.clj:359)
2024-06-11 15:56:56     at metabase.db.liquibase$run_in_scope_locked$reify__44374.run(liquibase.clj:324)
2024-06-11 15:56:56     at liquibase.Scope.lambda$child$0(scope.java:186)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:195)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:185)
2024-06-11 15:56:56     at liquibase.Scope.child(scope.java:164)
2024-06-11 15:56:56     at metabase.db.liquibase$run_in_scope_locked.invokeStatic(liquibase.clj:317)
2024-06-11 15:56:56     at metabase.db.liquibase$run_in_scope_locked.invoke(liquibase.clj:300)
2024-06-11 15:56:56     at metabase.db.liquibase$migrate_up_if_needed_BANG_.invokeStatic(liquibase.clj:348)
2024-06-11 15:56:56     at metabase.db.liquibase$migrate_up_if_needed_BANG_.invoke(liquibase.clj:341)
2024-06-11 15:56:56     at metabase.db.setup$migrate_BANG_$fn__53305.invoke(setup.clj:84)
2024-06-11 15:56:56     at metabase.db.liquibase$do_with_liquibase$f_STAR___44315.invoke(liquibase.clj:139)
2024-06-11 15:56:56     at metabase.db.liquibase$do_with_liquibase.invokeStatic(liquibase.clj:142)
2024-06-11 15:56:56     at metabase.db.liquibase$do_with_liquibase.invoke(liquibase.clj:130)
2024-06-11 15:56:56     at metabase.db.setup$migrate_BANG_.invokeStatic(setup.clj:73)
2024-06-11 15:56:56     at metabase.db.setup$migrate_BANG_.doInvoke(setup.clj:55)
2024-06-11 15:56:56     at clojure.lang.RestFn.invoke(restfn.java:425)
2024-06-11 15:56:56     at metabase.db.setup$run_schema_migrations_BANG_.invokeStatic(setup.clj:149)
2024-06-11 15:56:56     at metabase.db.setup$run_schema_migrations_BANG_.invoke(setup.clj:144)
2024-06-11 15:56:56     at metabase.db.setup$setup_db_BANG_$fn53333$fn53334.invoke(setup.clj:167)
2024-06-11 15:56:56     at metabase.util.jvm$do_with_us_locale.invokeStatic(jvm.clj:239)
2024-06-11 15:56:56     at metabase.util.jvm$do_with_us_locale.invoke(jvm.clj:225)
2024-06-11 15:56:56     at metabase.db.setup$setup_db_BANG_$fn__53333.invoke(setup.clj:161)
2024-06-11 15:56:56     at metabase.db.setup$setup_db_BANG_.invokeStatic(setup.clj:160)
2024-06-11 15:56:56     at metabase.db.setup$setup_db_BANG_.invoke(setup.clj:153)
2024-06-11 15:56:56     at metabase.db$setup_db_BANG_$fn__53358.invoke(db.clj:82)
2024-06-11 15:56:56     at metabase.db$setup_db_BANG_.invokeStatic(db.clj:77)
2024-06-11 15:56:56     at metabase.db$setup_db_BANG_.doInvoke(db.clj:64)
2024-06-11 15:56:56     at clojure.lang.RestFn.invoke(restfn.java:421)
2024-06-11 15:56:56     at metabase.core$init_BANG__STAR_.invokeStatic(core.clj:117)
2024-06-11 15:56:56     at metabase.core$init_BANG__STAR_.invoke(core.clj:98)
2024-06-11 15:56:56     at metabase.core$init_BANG_.invokeStatic(core.clj:170)
2024-06-11 15:56:56     at metabase.core$init_BANG_.invoke(core.clj:165)
2024-06-11 15:56:56     at metabase.core$start_normally.invokeStatic(core.clj:182)
2024-06-11 15:56:56     at metabase.core$start_normally.invoke(core.clj:176)
2024-06-11 15:56:56     at metabase.core$entrypoint.invokeStatic(core.clj:215)
2024-06-11 15:56:56     at metabase.core$entrypoint.doInvoke(core.clj:209)
2024-06-11 15:56:56     at clojure.lang.RestFn.invoke(restfn.java:397)
2024-06-11 15:56:56     at clojure.lang.AFn.applyToHelper(afn.java:152)
2024-06-11 15:56:56     at clojure.lang.RestFn.applyTo(restfn.java:132)
2024-06-11 15:56:56     at clojure.lang.Var.applyTo(var.java:705)
2024-06-11 15:56:56     at clojure.core$apply.invokeStatic(core.clj:667)
2024-06-11 15:56:56     at clojure.core$apply.invoke(core.clj:662)
2024-06-11 15:56:56     at metabase.bootstrap$_main.invokeStatic(bootstrap.clj:31)
2024-06-11 15:56:56     at metabase.bootstrap$_main.doInvoke(bootstrap.clj:28)
2024-06-11 15:56:56     at clojure.lang.RestFn.invoke(restfn.java:397)
2024-06-11 15:56:56     at clojure.lang.AFn.applyToHelper(afn.java:152)
2024-06-11 15:56:56     at clojure.lang.RestFn.applyTo(restfn.java:132)
2024-06-11 15:56:56     at metabase.bootstrap.main(Unknown Source)
2024-06-11 15:56:56 Caused by: org.postgresql.util.PSQLException: ERROR: invalid input syntax for type double precision: "BW2J+jGXv2seSOh2ZXEehWTUWTwUJx5lSUr20Sw6mulcSaaFxjEmtgtLwE+wacV5CqyebuY/zg/e5pP7Wg5m3Q=="
2024-06-11 15:56:56     at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(queryexecutorimpl.java:2725)
2024-06-11 15:56:56     at org.postgresql.core.v3.QueryExecutorImpl.processResults(queryexecutorimpl.java:2412)
2024-06-11 15:56:56     at org.postgresql.core.v3.QueryExecutorImpl.execute(queryexecutorimpl.java:371)
2024-06-11 15:56:56     at org.postgresql.jdbc.PgStatement.executeInternal(pgstatement.java:502)
2024-06-11 15:56:56     at org.postgresql.jdbc.PgStatement.execute(pgstatement.java:419)
2024-06-11 15:56:56     at org.postgresql.jdbc.PgStatement.executeWithFlags(pgstatement.java:341)
2024-06-11 15:56:56     at org.postgresql.jdbc.PgStatement.executeCachedSql(pgstatement.java:326)
2024-06-11 15:56:56     at org.postgresql.jdbc.PgStatement.executeWithFlags(pgstatement.java:302)
2024-06-11 15:56:56     at org.postgresql.jdbc.PgStatement.execute(pgstatement.java:297)
2024-06-11 15:56:56     at com.mchange.v2.c3p0.impl.NewProxyStatement.execute(newproxystatement.java:75)
2024-06-11 15:56:56     at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(jdbcexecutor.java:464)
2024-06-11 15:56:56     ... 86 more
2024-06-11 15:56:56 
2024-06-11 15:56:56 UPDATE SUMMARY
2024-06-11 15:56:56 Run:                         32
2024-06-11 15:56:56 Previously run:             302
2024-06-11 15:56:56 Filtered out:                49
2024-06-11 15:56:56 -------------------------------
2024-06-11 15:56:56 Total change sets:          383
```

### Information about your Metabase installation

```JSON
- Metabase version: 0.49.6 (tried the migration starting with the 0.49.13 as well
- Metabase environment: Docker
- Metabase internal database: PostgreSQL
```


### Severity

blocking usage of Metabase entirely

### Additional context

_No response_

--- a/resources/migrations/001_update_migrations.yaml
+++ b/resources/migrations/001_update_migrations.yaml
@@ -7227,25 +7227,6 @@ databaseChangeLog:
                   defaultValue: false
 
   - changeSet:
-      id: v50.2024-04-12T12:33:09
-      author: piranha
-      comment: 'Copy old cache configurations to cache_config table'
-      changes:
-        - sqlFile:
-            dbms: postgresql
-            path: custom_sql/fill_cache_config.pg.sql
-            relativeToChangelogFile: true
-        - sqlFile:
-            dbms: mysql,mariadb
-            path: custom_sql/fill_cache_config.my.sql
-            relativeToChangelogFile: true
-        - sqlFile:
-            dbms: h2
-            path: custom_sql/fill_cache_config.h2.sql
-            relativeToChangelogFile: true
-      rollback: # no rollback necessary, we're not removing the columns yet
-
-  - changeSet:
       id: v50.2024-04-15T16:30:35
       author: qnkhuat
       comment: Add report_card.last_used_at
@@ -7871,6 +7852,35 @@ databaseChangeLog:
             path: trash/mariadb.sql
             relativeToChangelogFile: true
       rollback: # not needed, columns will be deleted
+
+  - changeSet:
+      id: v50.2024-06-12T12:33:08
+      author: piranha
+      comment: 'Decrypt some settings so the next migration runs well'
+      changes:
+        - customChange:
+            class: "metabase.db.custom_migrations.DecryptCacheSettings"
+      rollback: # no rollback necessary, they'll be encrypted if you downgrade and touch them
+
+  - changeSet:
+      # this had id `v50.2024-04-12T12:33:09` before #44048
+      id: v50.2024-06-12T12:33:09
+      author: piranha
+      comment: 'Copy old cache configurations to cache_config table'
+      changes:
+        - sqlFile:
+            dbms: postgresql
+            path: custom_sql/fill_cache_config.pg.sql
+            relativeToChangelogFile: true
+        - sqlFile:
+            dbms: mysql,mariadb
+            path: custom_sql/fill_cache_config.my.sql
+            relativeToChangelogFile: true
+        - sqlFile:
+            dbms: h2
+            path: custom_sql/fill_cache_config.h2.sql
+            relativeToChangelogFile: true
+      rollback: # no rollback necessary, we're not removing the columns yet
 
   - changeSet:
       id: v51.2024-05-13T15:30:57

--- a/resources/migrations/001_update_migrations.yaml
+++ b/resources/migrations/001_update_migrations.yaml
@@ -7854,7 +7854,7 @@ databaseChangeLog:
       rollback: # not needed, columns will be deleted
 
   - changeSet:
-      id: v50.2024-06-12T12:33:08
+      id: v50.2024-06-12T12:33:07
       author: piranha
       comment: 'Decrypt some settings so the next migration runs well'
       changes:
@@ -7864,7 +7864,7 @@ databaseChangeLog:
 
   - changeSet:
       # this had id `v50.2024-04-12T12:33:09` before #44048
-      id: v50.2024-06-12T12:33:09
+      id: v50.2024-06-12T12:33:08
       author: piranha
       comment: 'Copy old cache configurations to cache_config table'
       changes:
@@ -7881,6 +7881,17 @@ databaseChangeLog:
             path: custom_sql/fill_cache_config.h2.sql
             relativeToChangelogFile: true
       rollback: # no rollback necessary, we're not removing the columns yet
+
+  - changeSet:
+      id: v50.2024-06-12T12:33:09
+      author: piranha
+      comment: 'Fix root cache config for mysql if it is wrong'
+      dbms: mysql,mariadb
+      changes:
+        - sqlFile:
+            path: custom_sql/fill_cache_config_root_fix.my.sql
+            relativeToChangelogFile: true
+      rollback: # no rollback necessary here too
 
   - changeSet:
       id: v51.2024-05-13T15:30:57

--- a/resources/migrations/custom_sql/fill_cache_config_root_fix.my.sql
+++ b/resources/migrations/custom_sql/fill_cache_config_root_fix.my.sql
@@ -1,0 +1,10 @@
+UPDATE cache_config
+   SET config = json_object(
+           'multiplier', coalesce((select cast(`value` as unsigned) from setting where `key` = 'query-caching-ttl-ratio'), 10),
+           'min_duration_ms', coalesce((select cast(`value` as unsigned) from setting where `key` = 'query-caching-min-ttl'), 60000)
+         )
+ WHERE model = 'root' AND
+       model_id = 0 AND
+       strategy = 'ttl' AND
+       (json_extract(config, '$.multiplier') = 0 OR
+        json_extract(config, '$.min_duration_ms') = 0);

--- a/src/metabase/db/custom_migrations.clj
+++ b/src/metabase/db/custom_migrations.clj
@@ -1458,3 +1458,10 @@
       (->> settings
            (filter :value)
            (t2/insert! :setting)))))
+
+(define-migration DecryptCacheSettings
+  (let [decrypt! (fn [k]
+                   (t2/update! :setting :key k {:value (raw-setting-value k)}))]
+    (run! decrypt! ["query-caching-ttl-ratio"
+                    "query-caching-min-ttl"
+                    "enable-query-caching"])))

--- a/test/metabase/db/custom_migrations_test.clj
+++ b/test/metabase/db/custom_migrations_test.clj
@@ -1998,19 +1998,19 @@
         (is (false? (sample-content-created?)))))))
 
 (deftest decrypt-cache-settings-test
-  (impl/test-migrations "v50.2024-06-12T12:33:08" [migrate!]
+  (impl/test-migrations "v50.2024-06-12T12:33:07" [migrate!]
     (encryption-test/with-secret-key "whateverwhatever"
       (mdb.query/update-or-insert! :model/Setting {:key "enable-query-caching"} (constantly {:value "true"}))
       (mdb.query/update-or-insert! :model/Setting {:key "query-caching-ttl-ratio"} (constantly {:value "100"}))
       (mdb.query/update-or-insert! :model/Setting {:key "query-caching-min-ttl"} (constantly {:value "123"})))
 
     (testing "Values were indeed encrypted"
-      (is (not= "true" (t2/select-one-fn :value :model/Setting :key "enable-query-caching"))))
+      (is (not= "true" (t2/select-one-fn :value :setting :key "enable-query-caching"))))
 
     (encryption-test/with-secret-key "whateverwhatever"
       (migrate!))
 
     (testing "But not anymore"
-      (is (= "true" (t2/select-one-fn :value :model/Setting :key "enable-query-caching")))
-      (is (= "100" (t2/select-one-fn :value :model/Setting :key "query-caching-ttl-ratio")))
-      (is (= "123" (t2/select-one-fn :value :model/Setting :key "query-caching-min-ttl"))))))
+      (is (= "true" (t2/select-one-fn :value :setting :key "enable-query-caching")))
+      (is (= "100" (t2/select-one-fn :value :setting :key "query-caching-ttl-ratio")))
+      (is (= "123" (t2/select-one-fn :value :setting :key "query-caching-min-ttl"))))))

--- a/test/metabase/db/custom_migrations_test.clj
+++ b/test/metabase/db/custom_migrations_test.clj
@@ -16,7 +16,6 @@
    [metabase.db :as mdb]
    [metabase.db.connection :as mdb.connection]
    [metabase.db.custom-migrations :as custom-migrations]
-   [metabase.db.query :as mdb.query]
    [metabase.db.schema-migrations-test.impl :as impl]
    [metabase.driver :as driver]
    [metabase.models.database :as database]
@@ -2000,9 +1999,9 @@
 (deftest decrypt-cache-settings-test
   (impl/test-migrations "v50.2024-06-12T12:33:07" [migrate!]
     (encryption-test/with-secret-key "whateverwhatever"
-      (mdb.query/update-or-insert! :model/Setting {:key "enable-query-caching"} (constantly {:value "true"}))
-      (mdb.query/update-or-insert! :model/Setting {:key "query-caching-ttl-ratio"} (constantly {:value "100"}))
-      (mdb.query/update-or-insert! :model/Setting {:key "query-caching-min-ttl"} (constantly {:value "123"})))
+      (t2/insert! :setting [{:key "enable-query-caching", :value (encryption/maybe-encrypt "true")}
+                            {:key "query-caching-ttl-ratio", :value (encryption/maybe-encrypt "100")}
+                            {:key "query-caching-min-ttl", :value (encryption/maybe-encrypt "123")}]))
 
     (testing "Values were indeed encrypted"
       (is (not= "true" (t2/select-one-fn :value :setting :key "enable-query-caching"))))

--- a/test/metabase/db/custom_migrations_test.clj
+++ b/test/metabase/db/custom_migrations_test.clj
@@ -16,6 +16,7 @@
    [metabase.db :as mdb]
    [metabase.db.connection :as mdb.connection]
    [metabase.db.custom-migrations :as custom-migrations]
+   [metabase.db.query :as mdb.query]
    [metabase.db.schema-migrations-test.impl :as impl]
    [metabase.driver :as driver]
    [metabase.models.database :as database]
@@ -1995,3 +1996,21 @@
           :date_joined   :%now})
         (migrate!)
         (is (false? (sample-content-created?)))))))
+
+(deftest decrypt-cache-settings-test
+  (impl/test-migrations "v50.2024-06-12T12:33:08" [migrate!]
+    (encryption-test/with-secret-key "whateverwhatever"
+      (mdb.query/update-or-insert! :model/Setting {:key "enable-query-caching"} (constantly {:value "true"}))
+      (mdb.query/update-or-insert! :model/Setting {:key "query-caching-ttl-ratio"} (constantly {:value "100"}))
+      (mdb.query/update-or-insert! :model/Setting {:key "query-caching-min-ttl"} (constantly {:value "123"})))
+
+    (testing "Values were indeed encrypted"
+      (is (not= "true" (t2/select-one-fn :value :model/Setting :key "enable-query-caching"))))
+
+    (encryption-test/with-secret-key "whateverwhatever"
+      (migrate!))
+
+    (testing "But not anymore"
+      (is (= "true" (t2/select-one-fn :value :model/Setting :key "enable-query-caching")))
+      (is (= "100" (t2/select-one-fn :value :model/Setting :key "query-caching-ttl-ratio")))
+      (is (= "123" (t2/select-one-fn :value :model/Setting :key "query-caching-min-ttl"))))))

--- a/test/metabase/db/schema_migrations_test.clj
+++ b/test/metabase/db/schema_migrations_test.clj
@@ -33,6 +33,7 @@
    [metabase.models.collection :as collection]
    [metabase.test :as mt]
    [metabase.test.fixtures :as fixtures]
+   [metabase.util.encryption-test :as encryption-test]
    [toucan2.core :as t2]))
 
 (set! *warn-on-reflection* true)
@@ -1494,10 +1495,13 @@
 
 (deftest cache-config-migration-test
   (testing "Caching config is correctly copied over"
-    (impl/test-migrations "v50.2024-04-12T12:33:09" [migrate!]
+    (impl/test-migrations ["v50.2024-06-12T12:33:08"] [migrate!]
+      ;; this peculiar setup is to reproduce #44012, `enable-query-caching` should be unencrypted for the condition
+      ;; to hit it
       (mdb.query/update-or-insert! :model/Setting {:key "enable-query-caching"} (constantly {:value "true"}))
-      (mdb.query/update-or-insert! :model/Setting {:key "query-caching-ttl-ratio"} (constantly {:value "100"}))
-      (mdb.query/update-or-insert! :model/Setting {:key "query-caching-min-ttl"} (constantly {:value "123"}))
+      (encryption-test/with-secret-key "whateverwhatever"
+        (mdb.query/update-or-insert! :model/Setting {:key "query-caching-ttl-ratio"} (constantly {:value "100"}))
+        (mdb.query/update-or-insert! :model/Setting {:key "query-caching-min-ttl"} (constantly {:value "123"})))
 
       (let [user (create-raw-user! (mt/random-email))
             db   (t2/insert-returning-pk! :metabase_database (-> (mt/with-temp-defaults Database)
@@ -1521,7 +1525,9 @@
                                            :database_id            db
                                            :created_at             :%now
                                            :updated_at             :%now})]
-        (migrate!)
+
+        (encryption-test/with-secret-key "whateverwhatever"
+          (migrate! :up))
 
         (is (=? [{:model    "root"
                   :model_id 0

--- a/test/metabase/db/schema_migrations_test.clj
+++ b/test/metabase/db/schema_migrations_test.clj
@@ -1495,7 +1495,7 @@
 
 (deftest cache-config-migration-test
   (testing "Caching config is correctly copied over"
-    (impl/test-migrations ["v50.2024-06-12T12:33:08"] [migrate!]
+    (impl/test-migrations ["v50.2024-06-12T12:33:07"] [migrate!]
       ;; this peculiar setup is to reproduce #44012, `enable-query-caching` should be unencrypted for the condition
       ;; to hit it
       (mdb.query/update-or-insert! :model/Setting {:key "enable-query-caching"} (constantly {:value "true"}))
@@ -1548,7 +1548,7 @@
                   :config   {:duration 30 :unit "hours"}}]
                 (t2/select :model/CacheConfig))))))
   (testing "And not copied if caching is disabled"
-    (impl/test-migrations "v50.2024-04-12T12:33:09" [migrate!]
+    (impl/test-migrations ["v50.2024-04-12T12:33:07"] [migrate!]
       (mdb.query/update-or-insert! :model/Setting {:key "enable-query-caching"} (constantly {:value "false"}))
       (mdb.query/update-or-insert! :model/Setting {:key "query-caching-ttl-ratio"} (constantly {:value "100"}))
       (mdb.query/update-or-insert! :model/Setting {:key "query-caching-min-ttl"} (constantly {:value "123"}))
@@ -1561,6 +1561,32 @@
       (migrate!)
       (is (= []
              (t2/select :model/CacheConfig))))))
+
+(deftest cache-config-mysql-update-test
+  (when (= (mdb/db-type) :mysql)
+    (testing "Root cache config for mysql is updated with correct values"
+      (impl/test-migrations ["v50.2024-06-12T12:33:07"] [migrate!]
+        (mdb.query/update-or-insert! :model/Setting {:key "enable-query-caching"} (constantly {:value "true"}))
+        (mdb.query/update-or-insert! :model/Setting {:key "query-caching-ttl-ratio"} (constantly {:value "100"}))
+        (mdb.query/update-or-insert! :model/Setting {:key "query-caching-min-ttl"} (constantly {:value "123"}))
+
+        ;; the idea here is that `v50.2024-04-12T12:33:09` during execution with partially encrypted data (see
+        ;; `cache-config-migration-test`) instead of throwing an error just silently put zeros in config. If config
+        ;; contains zeros, we assume human did not touch it yet and update with existing (decrypted thanks to
+        ;; `v50.2024-06-12T12:33:07`) settings
+        (mdb.query/update-or-insert! :model/CacheConfig {:model "root" :model_id 0}
+                                     (constantly {:strategy "ttl"
+                                                  :config   {:multiplier      0
+                                                             :min_duration_ms 0}}))
+
+        (migrate!)
+
+        (is (=? [{:model    "root"
+                  :model_id 0
+                  :strategy :ttl
+                  :config   {:multiplier      100
+                             :min_duration_ms 123}}]
+                (t2/select :model/CacheConfig)))))))
 
 (deftest split-data-permissions-migration-test
   (testing "View Data and Create Query permissions are created correctly based on existing data permissions"

--- a/test/metabase/db/schema_migrations_test.clj
+++ b/test/metabase/db/schema_migrations_test.clj
@@ -33,6 +33,7 @@
    [metabase.models.collection :as collection]
    [metabase.test :as mt]
    [metabase.test.fixtures :as fixtures]
+   [metabase.util.encryption :as encryption]
    [metabase.util.encryption-test :as encryption-test]
    [toucan2.core :as t2]))
 
@@ -1498,10 +1499,10 @@
     (impl/test-migrations ["v50.2024-06-12T12:33:07"] [migrate!]
       ;; this peculiar setup is to reproduce #44012, `enable-query-caching` should be unencrypted for the condition
       ;; to hit it
-      (mdb.query/update-or-insert! :model/Setting {:key "enable-query-caching"} (constantly {:value "true"}))
+      (t2/insert! :setting [{:key "enable-query-caching", :value (encryption/maybe-encrypt "true")}])
       (encryption-test/with-secret-key "whateverwhatever"
-        (mdb.query/update-or-insert! :model/Setting {:key "query-caching-ttl-ratio"} (constantly {:value "100"}))
-        (mdb.query/update-or-insert! :model/Setting {:key "query-caching-min-ttl"} (constantly {:value "123"})))
+        (t2/insert! :setting [{:key "query-caching-ttl-ratio", :value (encryption/maybe-encrypt "100")}
+                              {:key "query-caching-min-ttl", :value (encryption/maybe-encrypt "123")}]))
 
       (let [user (create-raw-user! (mt/random-email))
             db   (t2/insert-returning-pk! :metabase_database (-> (mt/with-temp-defaults Database)
@@ -1531,27 +1532,28 @@
 
         (is (=? [{:model    "root"
                   :model_id 0
-                  :strategy :ttl
+                  :strategy "ttl"
                   :config   {:multiplier      100
                              :min_duration_ms 123}}
                  {:model    "database"
                   :model_id db
-                  :strategy :duration
+                  :strategy "duration"
                   :config   {:duration 10 :unit "hours"}}
                  {:model    "dashboard"
                   :model_id dash
-                  :strategy :duration
+                  :strategy "duration"
                   :config   {:duration 20 :unit "hours"}}
                  {:model    "question"
                   :model_id card
-                  :strategy :duration
+                  :strategy "duration"
                   :config   {:duration 30 :unit "hours"}}]
-                (t2/select :model/CacheConfig))))))
+                (->> (t2/select :cache_config)
+                     (mapv #(update % :config json/decode true))))))))
   (testing "And not copied if caching is disabled"
     (impl/test-migrations ["v50.2024-04-12T12:33:07"] [migrate!]
-      (mdb.query/update-or-insert! :model/Setting {:key "enable-query-caching"} (constantly {:value "false"}))
-      (mdb.query/update-or-insert! :model/Setting {:key "query-caching-ttl-ratio"} (constantly {:value "100"}))
-      (mdb.query/update-or-insert! :model/Setting {:key "query-caching-min-ttl"} (constantly {:value "123"}))
+      (t2/insert! :setting [{:key "enable-query-caching", :value (encryption/maybe-encrypt "false")}
+                            {:key "query-caching-ttl-ratio", :value (encryption/maybe-encrypt "100")}
+                            {:key "query-caching-min-ttl", :value (encryption/maybe-encrypt "123")}])
 
       ;; this one to have custom configuration to check they are not copied over
       (t2/insert-returning-pk! :metabase_database (-> (mt/with-temp-defaults Database)
@@ -1560,33 +1562,35 @@
                                                       (assoc :cache_ttl 10)))
       (migrate!)
       (is (= []
-             (t2/select :model/CacheConfig))))))
+             (t2/select :cache_config))))))
 
 (deftest cache-config-mysql-update-test
   (when (= (mdb/db-type) :mysql)
     (testing "Root cache config for mysql is updated with correct values"
-      (impl/test-migrations ["v50.2024-06-12T12:33:07"] [migrate!]
-        (mdb.query/update-or-insert! :model/Setting {:key "enable-query-caching"} (constantly {:value "true"}))
-        (mdb.query/update-or-insert! :model/Setting {:key "query-caching-ttl-ratio"} (constantly {:value "100"}))
-        (mdb.query/update-or-insert! :model/Setting {:key "query-caching-min-ttl"} (constantly {:value "123"}))
+      (encryption-test/with-secret-key "whateverwhatever"
+        (impl/test-migrations ["v50.2024-06-12T12:33:07"] [migrate!]
+          (t2/insert! :setting [{:key "enable-query-caching", :value (encryption/maybe-encrypt "true")}
+                                {:key "query-caching-ttl-ratio", :value (encryption/maybe-encrypt "100")}
+                                {:key "query-caching-min-ttl", :value (encryption/maybe-encrypt "123")}])
 
-        ;; the idea here is that `v50.2024-04-12T12:33:09` during execution with partially encrypted data (see
-        ;; `cache-config-migration-test`) instead of throwing an error just silently put zeros in config. If config
-        ;; contains zeros, we assume human did not touch it yet and update with existing (decrypted thanks to
-        ;; `v50.2024-06-12T12:33:07`) settings
-        (mdb.query/update-or-insert! :model/CacheConfig {:model "root" :model_id 0}
-                                     (constantly {:strategy "ttl"
-                                                  :config   {:multiplier      0
-                                                             :min_duration_ms 0}}))
+          ;; the idea here is that `v50.2024-04-12T12:33:09` during execution with partially encrypted data (see
+          ;; `cache-config-migration-test`) instead of throwing an error just silently put zeros in config. If config
+          ;; contains zeros, we assume human did not touch it yet and update with existing (decrypted thanks to
+          ;; `v50.2024-06-12T12:33:07`) settings
+          (t2/insert! :cache_config {:model    "root"
+                                     :model_id 0
+                                     :strategy "ttl"
+                                     :config   (json/encode {:multiplier      0
+                                                             :min_duration_ms 0})})
+          (migrate!)
 
-        (migrate!)
-
-        (is (=? [{:model    "root"
-                  :model_id 0
-                  :strategy :ttl
-                  :config   {:multiplier      100
-                             :min_duration_ms 123}}]
-                (t2/select :model/CacheConfig)))))))
+          (is (=? {:model    "root"
+                   :model_id 0
+                   :strategy "ttl"
+                   :config {:multiplier      100
+                            :min_duration_ms 123}}
+                  (-> (t2/select-one :cache_config)
+                      (update :config json/decode true)))))))))
 
 (deftest split-data-permissions-migration-test
   (testing "View Data and Create Query permissions are created correctly based on existing data permissions"

